docs/core/extensions/httpclient-factory-keyed-di.md

```diff
@@ -0,0 +1,302 @@
+
+```csharp
+services.AddHttpClient("shared")
+    .AddAsKeyed(ServiceLifetime.Singleton) // explicit singleton
+    .UseSocketsHttpHandler((h, _) => h.PooledConnectionLifetime = TimeSpan.FromMinutes(2))
+    .SetHandlerLifetime(Timeout.InfiniteTimeSpan); // disable rotation
+services.AddSingleton<MySingleton>();
+
+public class MySingleton([FromKeyedServices("shared")] HttpClient shared) // { ...
+```
+
+### Beware of scope mismatch
+
+While Scoped lifetime is much less problematic for the Named `HttpClient`s (compared to Singleton and Transient pitfalls), it has its own catch.
+
+> [!IMPORTANT]
+> Keyed Scoped lifetime of a specific `HttpClient` instance is bound&mdash;as expected&mdash;to the "ordinary" application scope (for example, incoming request scope) where it was resolved from. However, it does NOT apply to the underlying message handler chain, which is still managed by the `IHttpClientFactory`, in the same way it is for the Named clients created directly from factory. `HttpClient`s with the _same_ name, but resolved (within a `HandlerLifetime` timeframe) in two different scopes (for example, two concurrent requests to the same endpoint), can reuse the _same_ `HttpMessageHandler` instance. That instance, in turn, has its own separate scope, as illustrated in the [Message handler scopes](httpclient-factory.md#message-handler-scopes-in-ihttpclientfactory).
+
+> [!NOTE]
+> The [Scope Mismatch](httpclient-factory-troubleshooting.md#httpclient-doesnt-respect-scoped-lifetime) problem is nasty and long-existing one, and as of .NET 9 still remains [unsolved](https://github.com/dotnet/runtime/issues/47091). From a service injected through the regular DI infra, you would expect all the dependencies to be satisfied from the same scope&mdash;but for the Keyed Scoped `HttpClient` instances, that's unfortunately not the case.
+
+## Keyed message handler chain
+
+For some advanced scenarios, you might want to access `HttpMessageHandler` chain directly, instead of an `HttpClient` object. `IHttpClientFactory` provides `IHttpMessageHandlerFactory` interface to create the handlers; and if you enable Keyed DI, then not only `HttpClient`, but also the respective `HttpMessageHandler` chain is registered as a Keyed service:
+
+```csharp
+services.AddHttpClient("keyed-handler").AddAsKeyed();
+
+var handler = provider.GetRequiredKeyedService<HttpMessageHandler>("keyed-handler");
+var invoker = new HttpMessageInvoker(handler, disposeHandler: false);
+```
+
+## How to: Switch from Typed approach to Keyed DI
+
+> [!NOTE]
+> We currently recommend using Keyed DI approach instead of Typed clients.
+
+A minimal-change switch from an existing Typed client to a Keyed dependency can look as follows:
+
+```diff
```

Comment:
For me the diff feels harder to parse than if we had two separate code blocks.

Reply:
I had the exactly opposite feeling tbh 😅
//in follow up